meta: edit collaborator nomination process (#57483)
nodejs/node: GOVERNANCE.md

### Nominating a new Collaborator

To nominate a new Collaborator:

1. **Optional but strongly recommended**: open a
   [discussion in the nodejs/collaborators][] repository. Provide a summary of
   the nominee's contributions (see below for an example).
2. **Optional but strongly recommended**: After sufficient wait time (e.g. 72
   hours), if the nomination proposal has received some support and no explicit
   block, add a comment in the private discussion stating you're planning on
   opening a public issue, e.g. "I see a number of approvals and no block, I'll
   be opening a public nomination issue if I don't hear any objections in the

>    next 72 hours".

Reviewer: I agree with @RafaelGSS. This is kind of confusing to me.

### Nominating a new Collaborator

To nominate a new Collaborator:

1. **Optional but strongly recommended**: open a

>    [discussion in the nodejs/collaborators][] repository. Provide a summary of
>    the nominee's contributions (see below for an example).

Reviewer 2: The collaborators team contains 86 people but only 49 are watching the collaborators repo, which is 57%, so a lot of them would not receive a notification in the feedback stage. So I would recommend pinging the team in the private discussion.
```suggestion
   [discussion in the nodejs/collaborators][] repository. Provide a summary of
   the nominee's contributions (see below for an example). Mention
   @nodejs/collaborators in the discussion to notify other collaborators about
   the nomination.
```

Author: I'm strongly -1 on that, if folks chose to not receive notifications, we should not force it – it will only encourage them to also unsubscribe from notifications to team mention

Reviewer 3: Perhaps we can add an optional-but-recommended subscription to the collaborators repo somewhere? For example, in https://github.com/nodejs/node/blob/36e89dd13cdf7d5a81e0d4b53c0070dd1790705e/onboarding.md?plain=1#L51

Reviewer 2: Mentioning it in the collaborator onboarding guide might be a nice solution: https://github.com/nodejs/node/pull/57527

> 2. **Optional but strongly recommended**: After sufficient wait time (e.g. 72

Reviewer 3: Nit: 'e.g.' might sound too vague and be interpreted as quite arbitrary value
```suggestion
2. **Optional but strongly recommended**: After sufficient wait time (at least 72
```
<details>

<summary>Also an absolute nit</summary>

Apparently the documentation style guide refers to the Microsoft Writing Style Guide here https://github.com/nodejs/node/blob/36e89dd13cdf7d5a81e0d4b53c0070dd1790705e/doc/README.md?plain=1#L226, which asks to avoid Latin abbreviations there https://learn.microsoft.com/en-us/style-guide/word-choice/use-us-spelling-avoid-non-english-words
But even the style guide itself uses `e.g.` twice, so it's probably not applicable. 😅 

</details>

Author: > might sound too vague and be interpreted as quite arbitrary value

I mean, it _is_ an arbitrary value, and since it's an optional step, I don't think it's useful to pretend it's a hard set deadline

> 2. **Optional but strongly recommended**: After sufficient wait time (e.g. 72

Reviewer: So previously we were waiting 7 days for private nomination and now we are changing it to 72 hours? Why this change?

   hours), if the nomination proposal has received some support and no explicit

>    block, add a comment in the private discussion stating you're planning on

Reviewer: This is also changing the nomination process in a way that requires I'm not comfortable with. If the nomination doesn't get enough attention because most collaborators are inactive, this will block node.js from introducing new collaborators.

If we are adding "if the nomination proposal has received some support" part, we should handle the part where there isn't enough support nor there isn't anybody blocking it.

The cost of explicitly defining these things, makes it easy to find a loophole or put yourself in a situation where you can not make a sane decision.

Author: It's an optional step, there's no way it could block the project

   opening a public issue, e.g. "I see a number of approvals and no block, I'll
   be opening a public nomination issue if I don't hear any objections in the
   next 72 hours".
3. **Optional but strongly recommended**: Privately contact the nominee to make

>    sure they're confortable with the nomination.

Reviewer 3: That's French. 😄 
```suggestion
   sure they're comfortable with the nomination.
```

4. Open an issue in the [nodejs/node][] repository. Provide a summary of

>    the nominee's contributions (see below for an example). Mention
>    @nodejs/collaborators in the issue to notify other collaborators about
>    the nomination.

Reviewer 2: The public ping can be made optional here if it has been done in the private discussion already:
```suggestion
   the nominee's contributions (see below for an example). Mention
   @nodejs/collaborators in the issue to notify other collaborators about
   the nomination if it hasn't already been done in the private discussion.
```

The _Optional but strongly recommended_ steps are optional in the sense that
skipping them would not invalidate the nomination, but it could put the nominee

> in a very awkward situation if a nomination they didn't ask for pops out of

Reviewer: Previous comment was mentioning "landing this nomination frictionless as possible" and right now, we're rewording it with "awkward" which is not the ideal way of describing this awkward situation.

nowhere only to be rejected. Do not skip those steps unless you're absolutely

> certain the nominee is fine with the public scrutiny.

Author: It's GHFM, it's nowhere to be found in the [spec](https://github.github.com/gfm/). Anyway I'd rather not, I prefer to keep it as a paragraph

   opening a public issue, e.g. "I see a number of approvals and no block, I'll
   be opening a public nomination issue if I don't hear any objections in the
   next 72 hours".

> 3. **Optional but strongly recommended**: Contact privately the nominee to make

Reviewer 4: ```suggestion
3. **Optional but strongly recommended**: Privately contact the nominee to make
```

  organization
* Other participation in the wider Node.js community

> If the private discussion receives some support (in form of explicit comments or

Reviewer: Can we make this explicit and say that at least one +1

```suggestion
If the private discussion receives at least one +1 from another collaborator (in form of explicit comments or
```

Author: I've decided to not try to define what support should be, I think the informal thing is enough given it's an optional step anyway. I'd expect the "I'm going to open the public issue in 72 hours" be enough if folks think there hasn't been enough support

Reviewer 5: I will rise it a bit (maybe +5 or +10?) so far we have 85 collaborators listed ([ref](https://github.com/nodejs/node#collaborators).

> thumbs-up reactions) and no explicit block after a few days, add a comment in the

Reviewer 6: If we can clarify here what constitutes an "explicit block" and handle the self-nomination loophole then I'm ok with the rest of these changes.

I don't believe a mandatory private nomination while allowing for self public nominations are compatible with each other.

Author: > I don't believe a mandatory private nomination while allowing for self public nominations are compatible with each other.

Why? The private nomination is to protect the nominee – because they, and the rest of the world, do not need to know they're being rejected if they haven't asked for anything. In the case of a self-nomination, we cannot hide from them whether they're being accepted, and they open themself to the risk of being rejected.

Reviewer 6: Because it can be easily gamed. If an existing collaborator does not want to follow the formal process, they can simply ask the person to nominate themselves and sidestep the formal process, causing issues for both other collaborators and the nominee who may not understand the possible risks. If we are making the private process mandatory then let's make it mandatory in all cases and we avoid this possibility.

Author: I’ll make another pass to the text to make the private discussion optional but strongly recommended, hopefully that would address this concern (FWIW I’m a -1 on removing self-nominations, having a private-only process could enable (not sure if it’s the right word) discriminatory gatekeeping)

Author: I've decided to not try to define what a block is, I think the informal thing is enough given it's an optional step anyway. I'd expect the "I'm going to open the public issue in 72 hours" be enough to let folks who are actually blocking to react in time in case their blocking comment wasn't explicit enough

> thumbs-up reactions) and no explicit block after a few days, add a comment in the

Reviewer 6: A "few days" can mean different things to different people. Let's make it explicitly one week (7 days).

private discussion stating you're planning on opening a public issue, e.g.
"I see a number of approvals and no block, I'll be opening a public
nomination issue if I don't hear any objections in the next 72 hours".
After sufficient time has passed, open an issue in the [nodejs/node][] repository.
Provide the same summary of the nominee's contributions.

> The nomination passes if no collaborators oppose it after one week. In the case

Reviewer: Does this still apply to public nomination, or both? If both, what's the purpose of private nomination?

Author: Gather objections before the public nomination is opened

Reviewer: If there are no objections on private nomination, can someone object on public nomination duration?

Author: Yes, if someone (collaborator or not) thinks a nomination should not proceed for whatever reason, they should obviously share their concerns – the goal of this PR is not to reduce the time when an objection can be made, it's to make it more likely it will happen before the public issue is opened.

Reviewer: Yes, but this doesn't disallow people from objecting in public nominations. It just legalizes private nomination part, previously it was not mandatory but recommended if nominator is interested in reducing friction. 

So my question is, does this change really help? If yes, to whom? To nominator? To nominee? To collaborators?

It seems like making it mandatory or not doesn't change anything, but add yet another step.

Author: I believe it does help yes, and I disagree it adds "yet another step", I don't think any nominations in the past 3 years did not start with a private discussion.

Author: Note that I’m saying it helps, not that it solves it.

> The nomination passes if no collaborators oppose it after one week. In the case

Reviewer 6: ```suggestion
 The nomination passes if no collaborators explicitly oppose the public nomination after one week. In the case
```

I would suggest we also include a definition of what explicit opposition means. Simply asking a clarifying question could be viewed as "opposition" by some and not by others.

Author: I've decided to not try to define what a block is, I think the informal thing is enough given it's an optional step anyway. I'd expect the "I'm going to open the public issue in 72 hours" be enough to let folks who are actually blocking to react in time in case their blocking comment wasn't explicit enough

  organization
* Other participation in the wider Node.js community

> Mention @nodejs/collaborators in the issue to notify other collaborators about

Reviewer 2: Could you please share the reasoning behind removing the requirement for at-mentioning the collaborators team? I tried going through the linked issue but I wasn't able to find which concern this change addresses.

Author: Just trying to reduce the number of notifications we're getting is all, also I don't think it's very useful, we don't use it in private discussions yet folks still participate in those

Author: Removing it from the governance doesn't mean we can no longer do it, it means the nomination doesn't "fail" if we forgot to at-mention the team. In any case, I'd recommend subscribing to nodejs/collaborators, that's where notifications are really important for reviewing nominations (before they become public), and it's not too noisy.

Reviewer 7: As I read things, the private issue is still optional. So if someone skips that step, and opens an issue on the main repo and doesn't tag me, I'll never hear about it.

I don't necessarily need to get notified about _both_ issues, but I'd like to get notified about one, ideally the private one. If we're going to change anything in this area, I would make the private issue steps mandatory and require at-notification there.

Author: > I would make the private issue steps mandatory

That's what this PR was doing, but I had to revert it, see https://github.com/nodejs/node/pull/57483#issuecomment-2726918949

Reviewer 7: Okay, so if that part is going to stay as is then can you revert this change too? I think it should be required that collaborators get notified about new nominations. It's not like nominations happen so often that it's a deluge.

Reviewer 2: Since removing the at-mention would make it harder for collaborators to notice the nomination, we should keep the at-mentioning text as is in this PR.

Adding the at-mention is easy. In the best case scenario, it keeps everyone informed. In the worst case, it only sends an extra notification. IMO the benefit of having everyone informed outweighs the inconvenience of an extra notification.

If you feel strongly about editing the at-mention part, I think separating this change into its own PR would make the current PR go in smoothly because the rest of the change is already an improvement which seems to have consensus.

Reviewer 8: I totally get the sentiment about respecting people's notification preferences. Buuut I know my GitHub notifications are not configured how I want them—it's somehow very difficult to get them right. I don't get notifications on my own repos despite "all" 🤯

So much as I would rather, I think this should stay 😞

I won't block, but I feel very strongly that at the current, this change should not be made.

  organization
* Other participation in the wider Node.js community

> The nomination passes if no collaborators oppose it after one week, and if the

Reviewer: I think we should re-add notifying collaborators mention. This is one of the requirements of each collaborator to follow up, and one of the topics that have a say in.

Author: It's still here

> nominee publicly accepts it. In the case

Reviewer: Why do we explicitly need to define "publicly accepts it"? What does this step solve/fix? What's the purpose?

Reviewer: There is also cases where we don't ask it publicly: https://github.com/nodejs/node/issues/55053.

I think explicitly adding this doesn't make a difference, and if it doesn't make a difference what's the goal of adding it?
